// zf2-cli/commands/create-model.js
var mustache = require('mustache')
  , fs       = require('fs');

module.exports = {
  generateModelClass: function(mod, name, attrs){
    attrs.unshift('id'); // always add the id ;)
    return render(mod, name, attrs, 'model');
  },
  generateTableGateway: function(mod, name, attrs){
    return render(mod, name, attrs, 'table_gateway');
  },
  addServiceManagerConfig: function(input, module, model){
    return input.split('\n').reduce(function(sum, line){
      if(line.indexOf('-- Table Gateways') > -1){
        line += '\n'+'use '+module+'\\Model\\'+model+'Table;';
      }else if (line.indexOf('-- factories') > -1){
        line += '\n'+render(module, model, [], 'Module_dbAdapter');
      }
      return sum + '\n' + line;
    });
  },
  execute:function(dir, module, model){
    // -- create both files
    // -- update Module.php
  }
};

function render(mod, name, attrs, template){
    var tpl = fs.readFileSync(__dirname+'/../templates/'+template+'.php').toString();
    return mustache.to_html(tpl, {
      module:mod,
      model: name,
      attrs:attrs,
      table:name.toLowerCase()
    });
}

// -- extract this...
function whitespace(str){
  var w = '';
  for(var i = 0; i < str.match(/\s/g).length;i++){
    w+= " ";
  }
  return w;
}
