// zf2-cli/commands/create-module.js
var async = require('async')
  , fs = require('fs')
  , mustache = require('mustache')
  , templates = __dirname + '/../templates/';

module.exports = function createModules(names, destination, cb){
  async.forEachSeries(names,function(name, cont){
    createModule(name, destination, function(){
      cb(null, name);
      cont();
    });
  });
};

function createModule(name, destination, cb){
  var m = dest(destination+'/module/'+name)
    , d = dest(destination);

  var dirs = [
        d('/module/'+name),
        m('/config'), 
        m('/src'), 
        m('/src/'+name), 
        m('/src/'+name+'/Controller'),
        m('/src/'+name+'/Form'),
        m('/src/'+name+'/Model'),
        m('/view'),
        m('/view/'+name.toLowerCase()),
        m('/view/'+name.toLowerCase()+'/'+name.toLowerCase())
   ]; 
   module.exports.writeToAppConfig(destination, name);
   async.forEachSeries(dirs, fs.mkdir, function(results){
      async.forEach(['Module.php', 'module.config.php', 'autoload_classmap.php'], function(file, cb){
        fs.readFile(templates+file, function(err, tpl){
          var absoluteFile = destination+'/module/'+name+'/config/'+file;
          fs.writeFile(absoluteFile, mustache.to_html(tpl.toString(), {name:name}), cb);
        });
      }, cb);
   });
}

function dest(destination){
  return function(location){
    return destination+location+'/';
  };
}
module.exports.writeToAppConfig = function(dir, moduleName){
  fs.readFile(dir+'/config/application.config.php', function(err, data){
    var newContent = module.exports.addToApplication(moduleName, data.toString());
    fs.writeFile(dir+'/config/application.config.php', newContent);
  });
}
module.exports.addToApplication = function(moduleName, appContent){
  return appContent.split('\n').reduce(function(sum, line){
    if(line.indexOf('Application') > -1){

      line += '\n'+whitespace(line)+"'"+moduleName+"',";
    }
    return sum+'\n'+line;
  });
};

function whitespace(str){
  var w = '';
  for(var i = 0; i < str.match(/\s/g).length;i++){
    w+= " ";
  }
  return w;
}
